### PR TITLE
Limit max pool size for parallel execution via config param

### PR DESCRIPTION
**Pull request by mpkorstanje**
## Overview

With Java 9+ the `ForkJoinPool` used by JUnit can be configured with a maximum pool size via the `junit.jupiter.execution.parallel.config.fixed.max-pool-size` property.

While the number of concurrently executing tests may exceed the configured parallelism when tests become blocked, it will not exceed the maximum pool size in these circumstances.

With the following configuration:

```properties
junit.jupiter.execution.parallel.enabled=true
junit.jupiter.execution.parallel.config.strategy=fixed
junit.jupiter.execution.parallel.mode.default=concurrent
junit.jupiter.execution.parallel.config.fixed.parallelism=2
```

This example will report between 2-5 tests running concurrently.

```java
class ExampleTest {

    private static final AtomicInteger running = new AtomicInteger();

    @BeforeEach
    void increment() {
        System.out.println("Running " + running.incrementAndGet());
    }

    @AfterEach
    void decrement() {
        running.decrementAndGet();
    }

    static IntStream numbers() {
        return IntStream.range(0, 1000);
    }

    @ParameterizedTest
    @MethodSource("numbers")
    void test(int i) throws ExecutionException, InterruptedException {
        Runnable sleep = () -> {
            try {
                Thread.sleep(600);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }
        };
        ForkJoinPool.commonPool().submit(sleep).get();
    }
}
```

By also configuring the `max-pool-size` we can ensure the concurrently executing test does not exceed the configured 2.

```properties
junit.jupiter.execution.parallel.config.fixed.max-pool-size=2
```

Additionally, because the `ForkJoinPool` will by default reject tasks that would exceed the maximum pool size the
`junit.jupiter.execution.parallel.config.fixed.saturate` property has been added and will default to `true`. There appears to be no reason to ever set this `false` but it is there should someone depend on the old behavior.

These changes were intentionally not made to the `dynamic` strategy to limit the scope of this pull request. While I can reasonably predict what behaviour users of the `fixed` strategy might expect, I can not say the same about the `dynamic` strategy.

Fixes: #2545
Fixes: #1858
Fixes: #3026

This PR replaces #3027.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by mpkorstanje** [comment]
@marcphilipp To update the changelog I'd need to know which mile stone would go into.

**Comment by sbrannen** [comment]
> To update the changelog I'd need to know which milestone would go into.

I just created [5.10 M1](https://github.com/junit-team/junit5/milestone/65). So feel free to use that.

If it ends up landing in a different milestone, we'll move the change log entry accordingly.

**Comment by sbrannen** [comment]
Note, however, that there is currently no file for the 5.10 M1 release notes.

So, you'll need to follow the instructions in `release-notes-TEMPLATE.adoc`.

If that seems too cumbersome, a core committer can create the new file for you.

**Comment by mpkorstanje** [comment]
Explanation of why this fixes the listed issues:

For #2545:

While the proposed solution involves a `ForkJoinWorkerThreadFactory` the underlying root cause follows from the inability to saturate the fork join pool. This was fixed by #2792 and can be achieved through a custom parallel strategy. This PR simplifies the process and should for this specific use case remove the need for more complicated solutions.

For #1858:

The root cause was explained in https://github.com/junit-team/junit5/issues/1858#issuecomment-482189645. When a test puts its executing thread to sleep, JUnit will start another test to keep up the desired parallism. However as explained in https://github.com/junit-team/junit5/issues/1858#issuecomment-482609294 when using Selenium threads are commonly put to sleep and so it is desirable to limit the maximum pool size to limit the maximum number of web drivers created.

This too was fixed by #2792 and can be achieved through a custom parallel strategy. This PR simplifies the process and should for this specific use case remove the need for more complicated solutions.

For #3026:

Essentially this issues requests a configuration parameter to limit the maximum pool size and allow saturation. Removing the need to create a custom class for this very common scenario,

**Comment by OPeyrusse** [comment]
Even though I am not a member of this repository, I would like to warn you that what you did will not work with people also using `ForkJoinPool`.
I explained in [this comment](https://github.com/junit-team/junit5/issues/1858#issuecomment-1281081183) on #1858 how a single thread can end up running 40+ tests "concurrently". By concurrently, I mean that 40+ tests can report as having start, before any can complete.
If this is a use case that you plan to support with this PR, I would gladly assist you in designing a test case that reproduce the issue. Don't hesitate to ping me if you want.

**Comment by mpkorstanje** [comment]
Thanks for the heads-up! I didn't see that specific problem explicitly mentioned in the issues before so I left it out of scope.

I reckon the current PR can still be salvaged by fixing the wording. 😅

**Comment by OPeyrusse** [comment]
@mpkorstanje Do you want me to create a special test and open an issue for that particular problem?

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 1]
```diff
@@ -2404,6 +2408,20 @@ The following table lists relevant properties for configuring parallel execution
 | a positive integer
 | no default value
 
+| ```junit.jupiter.execution.parallel.config.fixed.max-pool-size```
+| Desired maximum pool size of the underlying fork join pool for the ```fixed```
+  configuration strategy
+| a positive integer, must not be smaller then `junit.jupiter.execution.parallel.config.fixed.parallelism`
```
Comment (marcphilipp) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2414]:
```suggestion
| a positive integer, must not be less than `junit.jupiter.execution.parallel.config.fixed.parallelism`
```

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/Constants.java [hunk 2]
```diff
@@ -166,6 +168,40 @@ public final class Constants {
 	public static final String PARALLEL_CONFIG_FIXED_PARALLELISM_PROPERTY_NAME = PARALLEL_CONFIG_PREFIX
 			+ CONFIG_FIXED_PARALLELISM_PROPERTY_NAME;
 
+	/**
+	 * Property name used to configure the maximum pool size of the underlying
+	 * fork join pool for the {@code fixed} configuration strategy: {@value}
+	 *
+	 * <p>Value must be an integer and larger or equal to
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/Constants.java:175]:
```suggestion
	 * <p>Value must be an integer and greater than or equal to
```

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java [hunk 3]
```diff
@@ -114,6 +120,34 @@ public ParallelExecutionConfiguration createConfiguration(ConfigurationParameter
 	 */
 	public static final String CONFIG_FIXED_PARALLELISM_PROPERTY_NAME = "fixed.parallelism";
 
+	/**
+	 * Property name used to configure the maximum pool size of the underlying
+	 * fork join pool for the {@link #FIXED} configuration strategy.
+	 * <p>Value must be an integer and larger or equal to
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java:126]:
```suggestion
	 *
	 * <p>Value must be an integer and greater than or equal to
```

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java [hunk 4]
```diff
@@ -114,6 +120,34 @@ public ParallelExecutionConfiguration createConfiguration(ConfigurationParameter
...
+	 * @since 1.10
+	 * @see #FIXED
+	 */
+	@API(status = EXPERIMENTAL, since = "1.10")
+	public static final String CONFIG_FIXED_MAX_POOL_SIZE_PROPERTY_NAME = "fixed.max-pool-size";
+
+	/**
+	 * Property name used to enable saturation of the underlying fork join pool
+	 * for the {@link #FIXED} configuration strategy.
+	 * <p>When set to {@code false} the underlying fork join pool will reject
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java:139]:
```suggestion
	 *
	 * <p>When set to {@code false} the underlying fork join pool will reject
```

**Review by marcphilipp — APPROVED** [review]
Looks good except for a few minor documentation things.

**Comment by mpkorstanje** [comment]
@OPeyrusse yes, please do. I can't promise or commit to solving it. But perhaps someone can.

**Comment by mpkorstanje** [comment]
I think this is what @OPeyrusse had in mind. The code below will report between 2-6 active tests as the `.join()` call will end up starting other tests. The number of active threads however does not exceed 2. 

I've updated the documentation to say concurrent "threads" instead of "tests".

```java
class ExampleTest {
    private static final Runnable sleep = () -> {
        try {
            Thread.sleep(600);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    };

    private static final AtomicInteger running = new AtomicInteger();

    @BeforeEach
    void increment() {
        System.out.println("Running " + running.incrementAndGet());
    }

    @AfterEach
    void decrement() {
        running.decrementAndGet();
    }

    static IntStream numbers() {
        return IntStream.range(0, 1000);
    }

    @ParameterizedTest
    @MethodSource("numbers")
    void test(int i) {
        ForkJoinPool.commonPool()
                .submit(sleep)
                .join();
    }

    @ParameterizedTest
    @MethodSource("numbers")
    void test2(int i) {
        ForkJoinPool.commonPool()
                .submit(sleep)
                .join();
    }
}
```

**Comment by mpkorstanje** [comment]
@marcphilipp good to go.

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.10.0-M1.adoc [hunk 5]
```diff
@@ -23,7 +23,7 @@ GitHub.
 
 ==== New Features and Improvements
 
-* ❓
+* Support limiting the max-pool-size for parallel execution via a property
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.10.0-M1.adoc:26]:
```suggestion
* Support for limiting the `max-pool-size` for parallel execution via a configuration parameter
```

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.10.0-M1.adoc [hunk 6]
```diff
@@ -35,7 +35,15 @@ GitHub.
 
 ==== Deprecations and Breaking Changes
 
-* ❓
+* The `fixed` parallel execution strategy will allow the thread pool to be saturated by
+  default
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.10.0-M1.adoc:39]:
```suggestion
* The `fixed` parallel execution strategy now allows the thread pool to be saturated by
  default.
```

File: documentation/src/docs/asciidoc/release-notes/release-notes-5.10.0-M1.adoc [hunk 7]
```diff
@@ -35,7 +35,15 @@ GitHub.
...
+
+==== New Features and Improvements
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/release-notes/release-notes-5.10.0-M1.adoc:41]: This document now contains 2 `New Features and Improvements` sections.

Please remove the unused section.

File: documentation/src/docs/asciidoc/user-guide/writing-tests.adoc [hunk 8]
```diff
@@ -2404,6 +2408,20 @@ The following table lists relevant properties for configuring parallel execution
 | a positive integer
 | no default value
 
+| ```junit.jupiter.execution.parallel.config.fixed.max-pool-size```
+| Desired maximum pool size of the underlying fork join pool for the ```fixed```
```
Comment (sbrannen) [inline documentation/src/docs/asciidoc/user-guide/writing-tests.adoc:2412]:
```suggestion
| Desired maximum pool size of the underlying fork-join pool for the ```fixed```
```

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/Constants.java [hunk 9]
```diff
@@ -166,6 +168,40 @@ public final class Constants {
...
+	public static final String PARALLEL_CONFIG_FIXED_MAX_POOL_SIZE_PROPERTY_NAME = PARALLEL_CONFIG_PREFIX
+			+ CONFIG_FIXED_MAX_POOL_SIZE_PROPERTY_NAME;
+
+	/**
+	 * Property name used to enable saturation of the underlying fork join pool
+	 * for the {@code ffixed} configuration strategy: {@value}
+	 *
+	 * <p>When set to {@code false} the underlying fork join pool will reject
+	 * additional tasks when all available workers are busy and the maximum
+	 * pool-size would be exceeded.
```
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/Constants.java:193]:
```suggestion
	 * <p>When set to {@code false} the underlying fork-join pool will reject
	 * additional tasks if all available workers are busy and the maximum
	 * pool-size would be exceeded.
```

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java [hunk 10]
```diff
@@ -114,6 +120,36 @@ public ParallelExecutionConfiguration createConfiguration(ConfigurationParameter
 	 */
 	public static final String CONFIG_FIXED_PARALLELISM_PROPERTY_NAME = "fixed.parallelism";
 
+	/**
+	 * Property name used to configure the maximum pool size of the underlying
+	 * fork join pool for the {@link #FIXED} configuration strategy.
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java:125]:
```suggestion
	 * fork-join pool for the {@link #FIXED} configuration strategy.
```

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java [hunk 11]
```diff
@@ -114,6 +120,36 @@ public ParallelExecutionConfiguration createConfiguration(ConfigurationParameter
...
+	 * @see #FIXED
+	 */
+	@API(status = EXPERIMENTAL, since = "1.10")
+	public static final String CONFIG_FIXED_MAX_POOL_SIZE_PROPERTY_NAME = "fixed.max-pool-size";
+
+	/**
+	 * Property name used to enable saturation of the underlying fork join pool
+	 * for the {@link #FIXED} configuration strategy.
+	 *
+	 * <p>When set to {@code false} the underlying fork join pool will reject
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java:141]:
```suggestion
	 * <p>When set to {@code false} the underlying fork-join pool will reject
```

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java [hunk 12]
```diff
@@ -114,6 +120,36 @@ public ParallelExecutionConfiguration createConfiguration(ConfigurationParameter
...
+	 * additional tasks when all available workers are busy and the maximum
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/DefaultParallelExecutionConfigurationStrategy.java:142]:
```suggestion
	 * additional tasks if all available workers are busy and the maximum
```

**Review by sbrannen — CHANGES_REQUESTED** [review]
Thanks for the PR!

I've requested some minor changes to the documentation.

**Comment by mpkorstanje** [comment]
Changes applied! 

For fixing small mistakes and typos please do feel free to push directly to my branch.

**Review by sbrannen — APPROVED** [review]

**Pull request merged**